docs/sc-memory/api/cpp/extended/permissions_api.md

```diff
@@ -0,0 +1,809 @@
+```cpp
+ScMemoryContext context; // Some context with guest user.
+ScAddr const & guestUserAddr = context.GetUser();
+
+// Identify guest user.
+// Find a user, which was identified. You can provide your own logic of user
+// identification before.
+ScAddr const & userAddr1 = context.HelperFindBySystemIdtf("user_1");
+
+ScAddr const & arcAddr = ScMemory::ms_globalContext->CreateEdge(
+  ScType::EdgeDCommonConst, guestUserAddr, userAddr1);
+ScMemory::ms_globalContext->CreateEdge(
+  ScType::EdgeAccessConstPosPerm, ScKeynodes::nrel_identified_user, arcAddr);
+// Only `ScMemory::ms_globalContext` can identify users.
+```
+
+After, you can authenticate identified user.
+
+### **Example of user authentication**
+
+To do this programly you can create waiter to wait sc-event of adding outgoing sc-arc from `concept_authenticated_user` and request user to be authenticated, i.e. create sc-arc between `concept_authentication_request` and user. After waiting this sc-event, your user will be authenticated.
+
+```cpp
+...
+// Find a user, which you want to authenticate.
+ScAddr const & userAddr1 = context.HelperFindBySystemIdtf("user_1");
+
+// Before user authentication, specify that user isn't authenticated. 
+// This way, you can subscribe to sc-event of erasing negative sc-arc, 
+// because when you authenticate user, all existing negative sc-arcs 
+// between `concept_authenticated_user` and user are erased. This logic will 
+// help you to wait for user to be authenticated.
+ScMemory::ms_globalContext->CreateEdge(
+  ScType::EdgeAccessConstNegTemp, 
+  ScKeynodes::concept_authenticated_user,
+  userAddr1);
+
+// Create sc-event waiter to wait user to be authenticated.
+// You should subscribe to sc-event of erasing negative sc-arc incoming to
+// user.
+auto eventWaiter 
+  = ScMemory::ms_globalContext->CreateConditionWaiter<
+    ScEventEraseIncomingArc<ScType::EdgeAccessConstNegTemp>>(
+  userAddr1,
+  [&]() -> void
+  {
+    ScMemory::ms_globalContext->CreateEdge(
+      ScType::EdgeAccessConstPosPerm, 
+      ScKeynodes::concept_authentication_request, 
+      userAddr1);
+
+    // Only `ScMemory::ms_globalContext` can authenticate users.
+  },
+  [&](ScEventAddIncomingArc<ScType::EdgeAccessConstNegTemp> const & event) 
+    -> sc_bool
+  {
+    // Check that sc-arc from `concept_authenticated_user` is erased.
+    return event.GetArcSourceElement() 
+      == ScKeynodes::concept_authenticated_user;
+  });
+
+// After creation, call method `Wait` and specify time while you 
+// will wait sc-event for specified subscription sc-element.
+eventWaiter->Wait(200); // milliseconds
+// By default, this wait time equals to 5000 milliseconds.
+// You will wait until sc-event occurs or until specified time expires.
+```
+
+Yes, this way of waiting for user authentication is complicated, but it ensures that the NOT factors in the user information are accounted for. In the future, transaction mechanisms will be implemented in the sc-machine to simplify its API.
```

Comment:
How this flow help with NOT factors? We can have situation when we don't know that user was authenticated and don't that user wasn't authentificated

Reply:
Properly, there should be no such situation. The knowledge base should always capture knowledge of what the user's current state is

Reply:
Fix typo in ScEventAfterGenerateIncomingArc

Reply:
Fixed